crates/containerd-shim-wasmer/src/executor.rs

```diff
@@ -0,0 +1,118 @@
+
+        let args = oci::get_args(spec);
+        if args.is_empty() {
+            return Err(ExecutorError::InvalidArg);
+        }
+
+        self.start(spec, args)
+            .map_err(|err| ExecutorError::Other(format!("failed to start wasm: {}", err)))?;
+
+        std::process::exit(0)
+    }
+
+    fn can_handle(&self, spec: &containerd_shim_wasm::sandbox::oci::Spec) -> bool {
+        // check if the entrypoint of the spec is a wasm binary.
+        let (module_name, _method) = oci::get_module(spec);
+        let module_name = match module_name {
+            Some(m) => m,
+            None => {
+                log::info!("wasmer cannot handle this workload, no arguments provided");
+                return false;
+            }
+        };
+        let path = PathBuf::from(module_name);
+
+        // TODO: do we need to validate the wasm binary?
+        // ```rust
+        //   let bytes = std::fs::read(path).unwrap();
+        //   wasmparser::validate(&bytes).is_ok()
+        // ```
+
+        path.extension()
+            .map(|ext| ext.to_ascii_lowercase())
+            .is_some_and(|ext| ext == "wasm" || ext == "wat")
+    }
+}
+
+impl WasmerExecutor {
+    fn start(&self, spec: &Spec, args: &[String]) -> anyhow::Result<()> {
+        // already in the cgroup
+        let envs = oci_wasmer::env_to_wasi(spec);
+        log::info!("setting up wasi");
+
+        self.stdio.redirect()?;
+
+        log::info!("wasi context ready");
+        let (module_name, method) = oci::get_module(spec);
+        let module_name = match module_name {
+            Some(m) => m,
+            None => {
+                return Err(anyhow::format_err!(
+                    "no module provided, cannot load module from file within container"
+                ))
+            }
+        };
+
+        log::info!("loading module from file {} ", module_name);
+        let mut store = Store::new(self.engine.clone());
+        let module = Module::from_file(&store, module_name)?;
+
+        log::info!("Starting `tokio` runtime...");
+        let runtime = tokio::runtime::Builder::new_multi_thread()
+            .enable_all()
+            .build()
+            .unwrap();
+        let _guard = runtime.enter();
+
+        log::info!("Creating `WasiEnv`.... args: {:?}, envs: {:?}", args, envs);
+        let mut wasi_env = WasiEnv::builder(&method)
+            // unknown command: /wasi-demo-app.wasm
+            .args(args[1..].to_vec())
+            .envs(envs)
+            .finalize(&mut store)?;
+
+        log::info!("Instantiating module with WASI imports...");
+        let import_object = wasi_env.import_object(&mut store, &module)?;
+        let instance = Instance::new(&mut store, &module, &import_object)?;
+
+        wasi_env.initialize(&mut store, instance.clone())?;
+
+        log::info!("Call WASI `_start` function...");
+        let start_func = instance.exports.get_function(&method)?;
+        start_func.call(&mut store, &[])?;
```

Comment:
I also thought about building `WasiEnv` manually, but then decided to use `.run_with_store` because I had some problems with WASIX images. What advantage do you see in building `WasiEnv` manually?

Reply:
I have not tried your kind of method so far. I'm more trying to implement it using the official examples.

The `wasix image` issue is how it happens, maybe I'll be able to test it.

Reply:
> I have not tried your kind of method so far. I'm more trying to implement it using the official examples.

It's not really my method, it's just another example 😉: https://github.com/wasmerio/wasmer/blob/master/examples/wasi.rs#L44

I don't have a strong opinion about any of them, I'm just trying to understand the pros and cons.

> The `wasix image` issue is how it happens, maybe I'll be able to test it.

I just tested this example: https://github.com/wasix-org/wasix-rust-examples/tree/main/wasix-axum
With a `Dockerfile` like this:
```Dockerfile
FROM scratch
COPY --chmod=0755 target/wasm32-wasmer-wasi/debug/wasix-axum.wasm ./wasix-axum.wasm
ENTRYPOINT ["/wasix-axum.wasm"]
```
(To open a port you need to pass an env variable)